public-api/src/main/java/org/pmiops/workbench/google/IAPAuthService.java

```diff
@@ -0,0 +1,110 @@
+
+@Component
+public class IAPAuthService {
+
+    private static final String IAM_SCOPE = "https://www.googleapis.com/auth/iam";
+    private static final String OAUTH_TOKEN_URI = "https://www.googleapis.com/oauth2/v4/token";
+    private static final String JWT_BEARER_TOKEN_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer";
+    private static final long EXPIRATION_TIME_IN_MILLIS = 3600 * 1000L;
+    private static final HttpTransport httpTransport = new NetHttpTransport();
+
+    private final String clientId;
+    private final ServiceAccountCredentials credentials;
+    private DecodedJWT googleJwt;
+    private String token;
+
+    @Autowired
+    ResourceLoader resourceLoader;
+
+    public IAPAuthService() throws IOException {
+        this.clientId = "238501349883-965gu9qminos5dfcpusi43eokvd5i3io.apps.googleusercontent.com";
+        this.credentials = getCredentials();
+        synchronized (this) {
+            if (googleJwt == null || googleJwt.getExpiresAt().before(new Date())) {
+                googleJwt = getGoogleIdToken();
+            }
+            token = googleJwt.getToken();
+        }
+    }
+
+    public String getToken() {
+        return token;
+    }
+
+    public void setToken(String token) {
+        this.token = token;
+    }
+
+    private ServiceAccountCredentials getCredentials() throws IOException {
+        Resource resource = resourceLoader.getResource("classpath:test-circle-key.json");
+        GoogleCredentials credentials = null;
+        if (resource.exists()) {
+            File file = ResourceUtils.getFile("classpath:test-circle-key.json");
+            credentials = GoogleCredentials.fromStream(new FileInputStream(file));
+        }
+
+        // Service account credentials are required to sign the jwt token.
+        if (credentials == null || !(credentials instanceof ServiceAccountCredentials)) {
+            throw new RuntimeException("Google credentials : service accounts credentials expected");
+        }
+        return (ServiceAccountCredentials) credentials;
+    }
+
+    private DecodedJWT getGoogleIdToken() throws IOException {
+        String jwt = getSignedJwt();
+        final GenericData tokenRequest = new GenericData()
+                .set("grant_type", JWT_BEARER_TOKEN_GRANT_TYPE)
+                .set("assertion", jwt);
+        final UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+        final HttpRequestFactory requestFactory = httpTransport.createRequestFactory();
+
+        final HttpRequest request = requestFactory
+                .buildPostRequest(new GenericUrl(OAUTH_TOKEN_URI), content)
+                .setParser(new JsonObjectParser(JacksonFactory.getDefaultInstance()));
+
+        HttpResponse response = request.execute();
+        GenericData responseData = response.parseAs(GenericData.class);
+        String idToken = (String) responseData.get("id_token");
+        return JWT.decode(idToken);
+    }
+
+    private String getSignedJwt() {
```

Comment:
Is it not possible to use this code snippet from the documentation, instead of the explicit JWT signing? https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/iap/src/main/java/com/example/iap/BuildIapRequest.java